srml/contracts/src/wasm/runtime.rs

```diff
@@ -586,6 +586,86 @@ define_env!(Env, <E: Ext>,
 		Ok(())
 	},
 
+	// Attempt to restore the specified contract.
+	//
+	// This will compute a tombstone hash from the caller's storage and the given code hash and
+	// if the hash matches the hash found in the tombstone at the specified address - kill the
+	// caller contract and restore the destination contract and set the specified `rent_allowance`.
+	// All caller's funds are transfered to the destination.
+	//
+	// This function doesn't perform restoration right away but defers it to the end of the
+	// transaction. If there is no tombstone in the destination address or if the hashes don't match
+	// then restoration is cancelled and no changes are made.
+	//
+	// `dest_ptr`, `dest_len` - the pointer and the length of a buffer that encodes `T::AccountId`
+	// with the address of the to be restored contract.
+	// `code_hash_ptr`, `code_hash_len` - the pointer and the length of a buffer that encodes
+	// a code hash of the to be restored contract.
+	// `rent_allowance_ptr`, `rent_allowance_len` - the pointer and the length of a buffer that
+	// encodes the rent allowance that must be set in the case of successful restoration.
+	// `delta_ptr` is the pointer to the start of a buffer that has `delta_count` storage keys
+	// laid out sequentially.
+	ext_restore_to(
+		ctx,
+		dest_ptr: u32,
+		dest_len: u32,
+		code_hash_ptr: u32,
+		code_hash_len: u32,
+		rent_allowance_ptr: u32,
+		rent_allowance_len: u32,
+		delta_ptr: u32,
+		delta_count: u32
+	) => {
+		let dest = {
+			let dest_buf = read_sandbox_memory(ctx, dest_ptr, dest_len)?;
+			<<E as Ext>::T as system::Trait>::AccountId::decode(&mut &dest_buf[..])
+				.ok_or_else(|| sandbox::HostError)?
+		};
+		let code_hash = {
+			let code_hash_buf = read_sandbox_memory(ctx, code_hash_ptr, code_hash_len)?;
+			<CodeHash<<E as Ext>::T>>::decode(&mut &code_hash_buf[..])
+				.ok_or_else(|| sandbox::HostError)?
+		};
+		let rent_allowance = {
+			let rent_allowance_buf = read_sandbox_memory(
+				ctx,
+				rent_allowance_ptr,
+				rent_allowance_len
+			)?;
+			BalanceOf::<<E as Ext>::T>::decode(&mut &rent_allowance_buf[..])
+				.ok_or_else(|| sandbox::HostError)?
+		};
+		let delta = {
+			// We don't use `with_capacity` here to not eagerly allocate the user specified amount
+			// of memory.
+			let mut delta = Vec::new();
+			let mut key_ptr = delta_ptr;
+
+			for _ in 0..delta_count {
+				const KEY_SIZE: usize = 32;
+
+				// Read the delta into the provided buffer and collect it into the buffer.
+				let mut delta_key: StorageKey = [0; KEY_SIZE];
+				read_sandbox_memory_into_buf(ctx, key_ptr, &mut delta_key)?;
```

Comment:
Maybe not a good idea for the sake of API clarity, but if `delta` in `Ext::note_restore_to` was `Vec<[&u8]>` instead of `Vec<StorageKey>` you could read the entire vector in one call to `read_sandbox_memory_into_buf` which might be more efficient.

Reply:
`Vec<[&u8]>`? 

But yeah, the same as in https://github.com/paritytech/substrate/pull/3255#discussion_r308773284. I'd like to avoid optimizations like this for the moment preferring more clear API.